### PR TITLE
docs: Solutions architecture document

### PR DESCRIPTION
**Pull request by alexandre-allard**
**Component**: documentation

**Context**:

**Summary**:
This document describe what's a Solution and how the feature will be integrated in MetalK8s.

**Acceptance criteria**: 

---

See: #1993

### CODE REVIEW

**Comment by slaperche-scality** [comment]
This will be developed in 2.4? I thought it was for 2.5 (if 2.5, then the target branch is wrong).

**Comment by alexandre-allard** [comment]
> This will be developed in 2.4? I thought it was for 2.5 (if 2.5, then the target branch is wrong).

Hum, I'm not sure, since there is already some broken stuff done around Solutions in 2.4, I think at least the first iteration should target 2.4 (to fix what we have). But I may be wrong.

File: docs/developer/architecture/solutions.rst [hunk 1]
```diff
@@ -0,0 +1,240 @@
+Solutions
+=========
+
+Context
+-------
+
+As for now, if we want to deploy applications on a MetalK8s cluster,
+it's achievable by applying manifest through ``kubectl apply -f manifest.yaml``
+or using helm with charts.
```
Comment (slaperche-scality) [inline docs/developer/architecture/solutions.rst]: `Helm`

File: docs/developer/architecture/solutions.rst [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+Requirements
+------------
+
+* Everything must work offline.
+* Solutions must still be available after a reboot.
+* Check archives integrity and validity.
+* Deploy Solutions (operator & UI manifests).
+* Handle multiple Solutions with different versions.
+* Guidelines for Solutions developers.
+* Robustness against invalid format.
```
Comment (slaperche-scality) [inline docs/developer/architecture/solutions.rst]: Invalid format of what? You mean a corrupted ISO?

File: docs/developer/architecture/solutions.rst [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+* An Operator to deploy on the cluster
+* A UI for managing and monitoring the application
+
+Requirements
+------------
+
+* Everything must work offline.
+* Solutions must still be available after a reboot.
+* Check archives integrity and validity.
+* Deploy Solutions (operator & UI manifests).
```
Comment (slaperche-scality) [inline docs/developer/architecture/solutions.rst]: Anything about deploying some extra-YAML (such as StorageClass, multiple CRDs)?

File: docs/developer/architecture/solutions.rst (hunk 2 above)
Comment (alexandre-allard) [reply docs/developer/architecture/solutions.rst]: Yes, corrupted ISO, but also missing stuff in the archive such as the operator manifest or invalid manifest. It's a leftover (duplicate of Check archives integrity and validity). I'll remove it.

File: docs/developer/architecture/solutions.rst [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+To solve this, MetalK8s introduces the concept of Solutions.
+
+What is a Solution ?
+
+It's a packaged Kubernetes application, archived as an ISO disk image,
+containing:
+
+* A set of OCI images to inject in MetalK8s image registry
+* An Operator to deploy on the cluster
+* A UI for managing and monitoring the application
```
Comment (gdemonet) [inline docs/developer/architecture/solutions.rst]: Yes, this information should be present in this document, but it's more part of the design than the context IMO.

File: docs/developer/architecture/solutions.rst [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+Context
+-------
+
+As for now, if we want to deploy applications on a MetalK8s cluster,
+it's achievable by applying manifest through ``kubectl apply -f manifest.yaml``
+or using Helm_ with charts.
+
+These approaches work, but for an offline environment, the user must first
+pre-populate the MetalK8s registry with the container images listed in these
+manifests/charts.
```
Comment (gdemonet) [inline docs/developer/architecture/solutions.rst]: That's not something a user can do today - MetalK8s registry is not writable, so no one can add images there. The only way around this in offline environments is to manually inject the images you need on each `containerd`.

This point is interesting, and makes sense to write it as a requirement: "support offline deployment of applications with arbitrary container images"

File: docs/developer/architecture/solutions.rst [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+Moreover, there is no control on what's deployed, so it is difficult to
+enforce certain practices or provide tooling to ease deployment or
+lifecycle management of these applications.
```
Comment (gdemonet) [inline docs/developer/architecture/solutions.rst:18]: True as well, can be marked as a requirement

File: docs/developer/architecture/solutions.rst [hunk 7]
```diff
@@ -0,0 +1,241 @@
+Solutions
+=========
+
+Context
+-------
```
Comment (gdemonet) [inline docs/developer/architecture/solutions.rst:5]: Context should explain why we want MetalK8s to be responsible for deploying applications in the first place.

MetalK8s is meant as a foundation for a comprehensive product made of multiple applications, while maintaining K8s as "an implementation detail" for the end user. Offline constraints are also not well handled in the usual ecosystem (straight manifests or Helm charts).

From this context, we derive the set of requirements:
- handle offline
- low impact on the user (no deep K8s knowledge required)
- handle multi-versions deployments
...

File: docs/developer/architecture/solutions.rst [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+* A set of OCI images to inject in MetalK8s image registry
+* An Operator to deploy on the cluster
+* A UI for managing and monitoring the application
+
+.. _Helm: https://helm.sh/
+
+Requirements
+------------
+
+* Everything must work offline.
```
Comment (gdemonet) [inline docs/developer/architecture/solutions.rst]: Define "everything" (installation / upgrade / downgrade...)

File: docs/developer/architecture/solutions.rst [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+* Solutions must still be available after a reboot.
```
Comment (gdemonet) [inline docs/developer/architecture/solutions.rst]: Let's mention the word "Solution" only once in the design section, this will force us to better think about the split.

For this point specifically, we want applications to keep running after a reboot or a rescheduling of the containers (due to e.g. cluster member loss/drain).

File: docs/developer/architecture/solutions.rst [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+* Check archives integrity and validity.
```
Comment (gdemonet) [inline docs/developer/architecture/solutions.rst]: And authenticity as well (using signatures)

File: docs/developer/architecture/solutions.rst [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+* Deploy Solutions (operator, UI, CRDs, StorageClass, Grafana dashboards).
```
Comment (gdemonet) [inline docs/developer/architecture/solutions.rst]: That's design, not a requirement. What we need is **the ability to orchestrate the deployment and lifecycle of complex applications**.

The design, in turn, will consider this requirement, and say: "we can't encompass all kinds of _complex applications_ or necessary operations to manage their lifecycle - this is why we agree on a contract, embodied by the Operator pattern, that applications deployed in this framework **must** embed their operational knowledge in code for MetalK8s users to rely on".

File: docs/developer/architecture/solutions.rst [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+* Handle multiple Solutions with different versions.
+* Guidelines for Solutions developers.
```
Comment (gdemonet) [inline docs/developer/architecture/solutions.rst]: More than guidelines, we need a clear **contract** between MetalK8s and applications to deploy as part of this framework.

File: docs/developer/architecture/solutions.rst [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+~~~~~~~~~~~~~~~~~~~
+
+As a solution admin, I want to be able to deploy a Solution through its
+UI, without any specific knowledge about Kubernetes.
+
+Solution upgrade/downgrade
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+As a solution admin, I want to be able to easily upgrade or downgrade a
+Solution using its UI without any specific knowledge about Kubernetes.
```
Comment (gdemonet) [inline docs/developer/architecture/solutions.rst]: Both can be written as a single user story:

```
Application deployment and lifecycle
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

As an application administrator, I want to manage the deployment and lifecycle
(upgrade/downgrade/scaling/configuration) of an application using either a UI
or through simple CLI commands (both should be available).
```

File: docs/developer/architecture/solutions.rst [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+~~~~~~~~~~~~~~~
+
+As a cluster admin, I want to be able to import a Solution archive using a CLI
+tool, to make the Solution available for deployment.
+
+Environment creation/deletion/inspection
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+As a cluster admin, I want to be able to create/delete/inspect an Environment
+using the CLI.
```
Comment (gdemonet) [inline docs/developer/architecture/solutions.rst]: `SyntaxError: 'Environment' variable is not defined` :stuck_out_tongue:

File: docs/developer/architecture/solutions.rst [hunk 15]
```diff
@@ -0,0 +1,241 @@
...
+
+* Everything must work offline.
+* Solutions must still be available after a reboot.
+* Check archives integrity and validity.
+* Deploy Solutions (operator, UI, CRDs, StorageClass, Grafana dashboards).
+* Handle multiple Solutions with different versions.
+* Guidelines for Solutions developers.
+
+User Stories
+------------
```
Comment (gdemonet) [inline docs/developer/architecture/solutions.rst:41]: I may be wrong, but again here, I think the user stories would better be explained without assumptions on the design. For most of them, it's just about wording :)

File: docs/developer/architecture/solutions.rst [hunk 16]
```diff
@@ -0,0 +1,241 @@
...
+to the cluster, from the bootstrap node.
+
+Here is how it will look::
+
+    apiVersion: metalk8s.scality.com/v1alpha1
+    kind: SolutionsConfiguration
+    solutions:
+      - /solutions/storage_1.0.0.iso
+      - /solutions/storage_latest.iso
+      - /other_solutions/computing.iso
```
Comment (gdemonet) [inline docs/developer/architecture/solutions.rst]: This has changed to include an `active` version (mapping to the `stored` version of CRDs) - see the in-flight PR for the new format

File: docs/developer/architecture/solutions.rst [hunk 17]
```diff
@@ -0,0 +1,241 @@
...
+
+There is no explicit information about what an archive contains.
+Instead, we want the archive itself to contain such information and to
+discover it at import time.
```
Comment (gdemonet) [inline docs/developer/architecture/solutions.rst]: Even though this comment appears in the "SolutionsConfig" description, it reads like there will never be any assumption on the archive contents, which is wrong. Maybe reword to:

```
In this configuration file, no explicit information about the contents of archives
should appear. When read by Salt at import time, the archive metadata will be
discovered from the archive itself using a well-known file (`/product.txt`).
```

File: docs/developer/architecture/solutions.rst [hunk 18]
```diff
@@ -0,0 +1,241 @@
...
+
+This configuration file will be read by a Salt external pillar module,
+which will permit the consumption of it by Salt modules and states.
```
Comment (gdemonet) [inline docs/developer/architecture/solutions.rst]: Yes - let's add that it does the same as `BootstrapConfig` (file in `/etc/metalk8s`, mounted in the Salt master container...).

Let's also document the entire `metalk8s:solutions` pillar section in this design: what each key represents and how it's going to be computed.

File: docs/developer/architecture/solutions.rst [hunk 19]
```diff
@@ -0,0 +1,241 @@
...
+  (bytes) in length
+- Files are allowed to consist of multiple sections
+
+The generated archive should specify a volume ID, set to
+``{project_name} {version}``.
+
+.. _`ISO-9660:1988`: https://www.iso.org/obp/ui/#iso:std:iso:9660:ed-1:v1:en
+.. _`Rock Ridge`: https://en.wikipedia.org/wiki/Rock_Ridge
+.. _Joliet: https://en.wikipedia.org/wiki/Joliet_(file_system)
+.. _UTF-8: https://tools.ietf.org/html/rfc3629
```
Comment (gdemonet) [inline docs/developer/architecture/solutions.rst]: Let's leave this in the detailed guidelines / contract definition - just mention we choose ISO instead of tarballs, and explain why.

File: docs/developer/architecture/solutions.rst [hunk 20]
```diff
@@ -0,0 +1,241 @@
...
+
+OCI Images registry
+~~~~~~~~~~~~~~~~~~~
+
+Every container images from Solution archive will be exposed as a single
+repository through MetalK8s registry. The name of this repository will be
+computed from the product information ``<NAME>-<VERSION>``.
```
Comment (gdemonet) [inline docs/developer/architecture/solutions.rst:233]: We may need to handle name collisions (in case `<VERSION>` is not unique...). To manage this, we will need to inject this information (where an OCI repository sits for a given version of a Solution) into the main components (UI, Operator) of each Solution instance.

File: docs/developer/architecture/solutions.rst [hunk 21]
```diff
@@ -0,0 +1,241 @@
...
+
+Solution environment
+~~~~~~~~~~~~~~~~~~~~
+
+Solutions will be deployed into an ``Environment``, which is basically a
+namespace or a group of namespaces with a specific label
+``solutions.metalk8s.scality.com/environment``. It allows to run multiple
+instances of a Solution, with different versions, on the same cluster, without
+collision between them.
```
Comment (gdemonet) [inline docs/developer/architecture/solutions.rst]: Good point - in the user stories, add one mentioning the need to deploy new versions on "test" workloads before impacting "prod" services - and correlate this story with the design of "Environments".

File: docs/developer/architecture/solutions.rst [hunk 22]
```diff
@@ -0,0 +1,241 @@
...
+
+Solution management
+~~~~~~~~~~~~~~~~~~~
+
+We will provide CLI and UI to import, unimport and handle the whole lifecycle
+of a Solution. These tools are wrapper around Salt formulas.
```
Comment (gdemonet) [inline docs/developer/architecture/solutions.rst]: "whole lifecycle of a Solution" - yes, but can we show a simplified diagram to readers? The full UML diagram is really precise on how Salt should work to implement those "commands", but users of the feature could benefit from a higher-level schema.

File: docs/developer/architecture/solutions.rst [hunk 23]
```diff
@@ -0,0 +1,241 @@
...
+.. todo:: Try to details more / see if there is missing stuff.
+
+Iteration 1
+~~~~~~~~~~~
+
+* Solution example.
+* Tooling to import/unimport Solution (simple shell script).
+* Salt formulas to manage Solution (deployment and lifecycle).
+* MetalK8s UI to manage Solution.
+* Solution automated tests (deployment, upgrade/downgrade, deletion, ...).
```
Comment (gdemonet) [inline docs/developer/architecture/solutions.rst]: OK, but post-merge please :)

File: docs/developer/architecture/solutions.rst [hunk 24]
```diff
@@ -0,0 +1,241 @@
...
+
+Implementation Details
+----------------------
+
+.. todo:: Try to details more / see if there is missing stuff.
+
+Iteration 1
+~~~~~~~~~~~
+
+* Solution example.
```
Comment (gdemonet) [inline docs/developer/architecture/solutions.rst]: Mention that this example is "fake" - no real application deployed, simply used for testing Metal tooling.

File: docs/developer/architecture/solutions.rst [hunk 25]
```diff
@@ -0,0 +1,241 @@
...
+* MetalK8s UI to manage Solution.
+* Solution automated tests (deployment, upgrade/downgrade, deletion, ...).
+
+Iteration 2
+~~~~~~~~~~~
+
+* MetalK8s CLI to manage Solutions (supersedes import script & wraps Salt
+  call).
+* Integration into monitoring tools (Grafana dashboards, Alerting, ...).
+* Integration with the identity provider (Dex).
```
Comment (gdemonet) [inline docs/developer/architecture/solutions.rst:355]: Yes, but those are mainly documentation IMO - could be good to state this in the doc.

File: docs/developer/architecture/solutions.rst [hunk 26]
```diff
@@ -0,0 +1,241 @@
...
+
+In the Developer Guide:
+
+* Document how to monitor a Solution (ServiceMonitor, Service, ...).
+* Document how to build a Solution (layout, how to package, ...).
+
+Test Plan
+---------
+
+First of all, we must develop a Solution example to be able to tests the whole
```
Comment (gdemonet) [inline docs/developer/architecture/solutions.rst]:
```suggestion
First of all, we must develop a Solution example to be able to test the whole
```

File: docs/developer/architecture/solutions.rst [hunk 27]
```diff
@@ -0,0 +1,241 @@
...
+We will provide CLI and UI to import, unimport and handle the whole lifecycle
+of a Solution. These tools are wrapper around Salt formulas.
+
+Interaction diagram
+~~~~~~~~~~~~~~~~~~~
+
+We include a detailed interaction sequence diagram for describing how MetalK8s
+will handle user input when deploying / upgrading Solutions.
+
+.. uml:: solutions.uml
```
Comment (gdemonet) [inline docs/developer/architecture/solutions.rst]: This looks outdated... ;)

**Review by gdemonet — CHANGES_REQUESTED** [review]
Early comments, will need to refine the design proposal to make it clear "why" we make those choices, and not only "what" the choices are.

**Review by ChengYanJin — COMMENTED** [review]
LGTM!
Only one point, for the moment, UI doesn't handle remove a solution from an Environment. 
I don't know if we need to add it to the docs though.

File: docs/developer/architecture/solutions.rst [hunk 28]
```diff
@@ -0,0 +1,388 @@
...
+    active:
+      solution-name: X.Y.Z-suffix (or 'latest')
+
+In this configuration file, no explicit information about the contents of
+archives should appear. When read by Salt at import time, the archive metadata
+will be discovered from the archive itself using a well-known file
+(``/product.txt``), which needs to define at least these 2 lines::
+
+    NAME=<solution-name>
+    VERSION=<solution-ersion>
```
Comment (gdemonet) [inline docs/developer/architecture/solutions.rst]:
```suggestion
    VERSION=<solution-version>
```

File: docs/developer/architecture/solutions.rst [hunk 29]
```diff
@@ -0,0 +1,388 @@
...
+
+It will also be possible to define specific configuration for a Solution, using
+a ``config.yaml`` file at the root of the archive, with the following format::
```
Comment (gdemonet) [inline docs/developer/architecture/solutions.rst:126]: OK, but that's a bit redundant with the `product.txt`, no? Maybe it'd make sense to only request the more flexible YAML one (and also do it for our own ISO)?

File: docs/developer/architecture/solutions.rst [hunk 30]
```diff
@@ -0,0 +1,388 @@
...
+    operator:
+      image:
+        name: solution-name-operator
+        tag: 1.0.0
+    ui:
+      image:
+        name: solution-name-ui
+        tag: 1.0.0
+    customApiGroups:
+      - custom-api-group
```
Comment (gdemonet) [inline docs/developer/architecture/solutions.rst:139]: This will change - maybe add a `.. todo::` linking to #2372

File: docs/developer/architecture/solutions.rst [hunk 31]
```diff
@@ -0,0 +1,388 @@
...
+                  solution-name: 1.0.0
+              solution-b-namespace:
+                config: {}
+
+Archive format
+~~~~~~~~~~~~~~
+
+The archive will be packaged as an ISO image.
+
+We choosed the ISO image format instead of a compressed archive,
```
Comment (gdemonet) [inline docs/developer/architecture/solutions.rst]:
```suggestion
We chose the ISO image format instead of a compressed archive,
```

File: docs/developer/architecture/solutions.rst [hunk 32]
```diff
@@ -0,0 +1,388 @@
...
+    apiVersion: v1
+    kind: Namespace
+    metadata:
+      annotations:
+        solutions.metalk8s.scality.com/environment-description: <env-description>
+      labels:
+        solutions.metalk8s.scality.com/environment: <env-name>
+      name: <namespace-name>
+
+It allows to run multiple instances of a Solution, with different versions,
```
Comment (gdemonet) [inline docs/developer/architecture/solutions.rst]:
```suggestion
It allows to run multiple instances of a Solution, optionally with different versions,
```

File: docs/developer/architecture/solutions.rst [hunk 33]
```diff
@@ -0,0 +1,388 @@
...
+* Solution example, this is a fake application, with no other goal than
+  allowing testing of MetalK8s Solutions tooling.
+* Salt formulas to manage Solution (deployment and lifecycle).
+* Tooling around Salt formulas to ease Solutions management
+  (simple shell script).
+* MetalK8s UI to manage Solution.
+* Solution automated tests (deployment, upgrade/downgrade, deletion, ...)
+  in post-merge.
+
+Iteration 2
```
Comment (gdemonet) [inline docs/developer/architecture/solutions.rst:349]: Maybe add something about #2372 here as well

File: docs/developer/architecture/solutions.rst [hunk 34]
```diff
@@ -0,0 +1,388 @@
...
+
+* Document how to monitor a Solution (ServiceMonitor, Service, ...).
+* Document how to interface with the identity provider (Dex).
+* Document how to build a Solution (layout, how to package, ...).
+
+Test Plan
+---------
+
+First of all, we must develop a Solution example to be able to test the whole
+feature.
```
Comment (gdemonet) [inline docs/developer/architecture/solutions.rst]: We even need at least two versions of this example, to test upgrade/downgrade.

File: docs/developer/architecture/solutions.rst [hunk 35]
```diff
@@ -0,0 +1,388 @@
...
+
+   * Importing / removing a Solution archive
+   * Activating / deactivating a Solution
+   * Creating / deleting an Environment
+   * Adding / removing a Solution in / from an Environment
+   * Upgrading / downgrading a Solution
+
+* Solution can be plugged to MetalK8s cluster services
+  (monitoring, alerting, ...).
+* Solution can be managed through its UI.
```
Comment (gdemonet) [inline docs/developer/architecture/solutions.rst]: > its UI

Which UI? Solution UI? Or Metal UI?

**Review by gdemonet — CHANGES_REQUESTED** [review]
Minor changes, should be merge-able today :)

File: docs/developer/architecture/solutions.rst (hunk 29 above)
Comment (alexandre-allard) [reply docs/developer/architecture/solutions.rst:126]: Yes, I agree but that does not reflect what we have now, If we want to share this architecture document with Solutions developers, it'll be confusing.
Maybe we can add a todo with the linked issue as suggested below and do this in iteration 2.

File: docs/developer/architecture/solutions.rst (hunk 30 above)
Comment (alexandre-allard) [reply docs/developer/architecture/solutions.rst:139]: You mean #2389, but yes, will do.

File: docs/developer/architecture/solutions.rst (hunk 29 above)
Comment (alexandre-allard) [reply docs/developer/architecture/solutions.rst:126]: I've created #2422 to track this improvement.

Comment (gdemonet) [reply docs/developer/architecture/solutions.rst:126]: Arch docs are not developer docs - we are not really "sharing" them for people to act on, just to explain and describe our features and how we **plan** to implement them. Shortcuts (like a partial implementation or currently incomplete) could also be documented here, but it's more for us than for the outside.
On this aspect, I think you could have put less details in this document for the exact implementation (like format of pillar/configmaps...). Now that they are here, let's keep them, but just wanted to point that out.

But yeah, another ticket is perfectly fine to track this topic, as we don't know exactly what action to take yet.

File: docs/developer/architecture/solutions.rst (hunk 30 above)
Comment (gdemonet) [reply docs/developer/architecture/solutions.rst:139]: Right :)

**Review by gdemonet — APPROVED** [review]
:shipit:
